### PR TITLE
Reorder trait bound modifiers *after* `for<...>` binder in trait bounds

### PR DESCRIPTION
**Pull request by compiler-errors**
This PR suggests changing the grammar of trait bounds from:

```
[CONSTNESS] [ASYNCNESS] [?] [BINDER] [TRAIT_PATH]

const async ? for<'a> Sized
```

to

```
([BINDER] [CONSTNESS] [ASYNCNESS] | [?]) [TRAIT_PATH]
```

i.e., either

```
? Sized
```

or 

```
for<'a> const async Sized
```

(but not both)

### Why?

I think it's strange that the binder applies "more tightly" than the `?` trait polarity. This becomes even weirder when considering that we (or at least, I) want to have `async` trait bounds expressed like:

```
where T: for<'a> async Fn(&'a ()) -> i32,
```

and not:

```
where T: async for<'a> Fn(&'a ()) -> i32,
```

### Fallout

No crates on crater use this syntax, presumably because it's literally useless. This will require modifying the reference grammar, though.

### Alternatives

If this is not desirable, then we can alternatively keep parsing `for<'a>` after the `?` but deprecate it with either an FCW (or an immediate hard error), and begin parsing `for<'a>` *before* the `?`.

### CODE REVIEW

File: compiler/rustc_parse/src/parser/ty.rs
Comment (compiler-errors) [inline compiler/rustc_parse/src/parser/ty.rs:1006]: We could also just make this into a warning 🤔

**Comment by compiler-errors** [comment]
@craterbot check

File: compiler/rustc_parse/src/parser/ty.rs
Comment (fmease) [inline compiler/rustc_parse/src/parser/ty.rs:839]: Bounds like `?use<>`, `async use<>`, `const use<>` are no longer syntactically well-formed. That's actually great because `fn f() -> impl Sized + const use<> {}` (…) passes compilation on nightly, lol^^.

Seems like this warrants a regression test.

File: compiler/rustc_parse/src/parser/ty.rs
Comment (fmease) [inline compiler/rustc_parse/src/parser/ty.rs:864]: :pray:

**Review by fmease — APPROVED** [review]
r=me with nits addressed and after a lang FCP once the crater results come back (assuming they don't paint a grim picture)

File: compiler/rustc_parse/src/parser/ty.rs
Comment (compiler-errors) [reply compiler/rustc_parse/src/parser/ty.rs:839]: ugh i will fix `?use<>` lmao

**Comment by compiler-errors** [comment]
@craterbot check crates=https://crater-reports.s3.amazonaws.com/pr-127054/retry-regressed-list.txt

**Comment by traviscross** [comment]
@rfcbot fcp merge

We discussed this in the triage call today.  The specific proposal on the table here is doing what CE proposed above and tested via crater run: changing this to what we want via a straightforwardly breaking change.  We're comfortable doing this based on the observed lack of fallout and the unlikeliness that anyone would have had a reason to use this.

**Comment by tmandry** [comment]
Having thought about it more and talked with @compiler-errors, I would advocate for going even further than the PR does, and changing the grammar to this:

```
(? | [BINDER] [CONSTNESS] [ASYNCNESS]) [TRAIT_PATH]

? Sized
OR
for<'a> const async Sized
```

This prevents us from adding the confusing case of `for<'a> ?Sized` to the grammar. It's not clear what that would mean: "For all lifetimes `'a`, I don't necessarily implement this bound" is not really coherent. I think should really be `?for<'a> Sized` (which the grammar supports now) but I am fine taking that out altogether, since no one uses it and `?` bounds are confusing enough.

**Comment by nikomatsakis** [comment]
@rfcbot reviewed

I agree with the "OP"

**Comment by joshtriplett** [comment]
@tmandry Zero objection to making them entirely mutually exclusive.

**Comment by nikomatsakis** [comment]
Sorry, I was confused in my comment here. My opinion is that the grammar should be

```
Bound = Binder? Mode? TraitName Parameters?

Binder = `for<'a, 'b>`

Mode = `?` 

TraitName = `async` TraitName | Id

Parameters = `<` GenericArg `>`
```

**Comment by compiler-errors** [comment]
@nikomatsakis: If you're ok w `?` and `BINDER` being mutually exclusive, then let's just do that for now. 

I agree with @nikomatsakis (and consequently disagree with @tmandry) in that the binder should be outermost part of the trait bound, but let's not block this on that since it's really kinda useless syntax for now.

**Comment by traviscross** [comment]
@rfcbot reviewed

The grammar proposed by Niko [here](https://github.com/rust-lang/rust/pull/127054#issuecomment-2221478061) sounds right to me.  +1 on that.

I'm also +1 on making `?` and `BINDER` mutually exclusive for now to unblock this.

**Comment by tmandry** [comment]
Looking at this again I think there's an argument to be made for the `?for<'a> Trait`, but I also think that it doesn't really matter because they're both modifiers on a single trait. So regardless of what order you choose it can only mean one thing. I also agree that `for<'a> ?Trait` is a lot nicer looking. But even conceding that would leave the question of `?async Trait` or `async ?Trait`. So I still would prefer to disallow the combination for now and to revisit this in the future if we need to.

**Comment by tmandry** [comment]
@rfcbot reviewed
@rfcbot concern disallow combining `?` with binders and modifiers

**Comment by compiler-errors** [comment]
OK, we should now be enforcing that `?` is mutually exclusive with `for<...>` and trait bound modifiers (`async`/`const`)

**Comment by tmandry** [comment]
@rfcbot resolve disallow combining `?` with binders and modifiers

File: compiler/rustc_parse/src/parser/ty.rs
Comment (fmease) [inline compiler/rustc_parse/src/parser/ty.rs:1010]:
```suggestion
    /// TY_BOUND_NOPAREN = [for<GENERIC_PARAMS> CONSTNESS ASYNCNESS | POLARITY] SIMPLE_PATH
```

File: compiler/rustc_parse/src/parser/ty.rs
Comment (fmease) [inline compiler/rustc_parse/src/parser/ty.rs:931]:
````suggestion
    /// CONSTNESS = [["~"] "const"]
    /// ASYNCNESS = ["async"]
    /// POLARITY = ["?" | "!"]
    /// ```
    ///
    /// See [`Self::parse_generic_ty_bound`] for the complete grammar of trait bound modifiers.
````

Comment (fmease) [reply compiler/rustc_parse/src/parser/ty.rs]: Yuup, the discrepancy between the grammar snippets & the recursive descend parser (not semantically only organizationally) is giving me the ick

Comment (fmease) [reply compiler/rustc_parse/src/parser/ty.rs:931]: Also, the EBNF snippets in `rustc_parse` are like my worst nightmare lol, switching between different styles (EBNF is an open *family* of dialects anyway practically speaking) and being beyond outdated. I've been meaning to make everything consistent in one big PR for a while now, but er, not sure if it's worth my energy ^^'

**Review by fmease — APPROVED** [review]

**Comment by Kobzol** [comment]
@rust-timer build 0d5c969dd048d26e1488db4ecee869b4f484d396

**Pull request merged**
